[plsql] ParseException with SELECT INTO record defined as global variable

<!-- Please, prefix the report title with the language it applies to within brackets, such as [java] or [apex].
If not specific to a language, you can use [core]. -->

<!-- NB: issues about the rule designer should be opened at https://github.com/pmd/pmd-designer/issues -->

**Affects PMD Version:** 6.17.0

**Rule:**

**Description:** In below code sample I try to insert into t4666.rec.agent_no
t4666 is package
rec is record type
agent_no is records field

```
lip 23, 2019 2:56:17 PM net.sourceforge.pmd.RulesetsFactoryUtils printRuleNamesInDebug
FINER: Loaded rule AvoidTabCharacter
lip 23, 2019 2:56:17 PM net.sourceforge.pmd.PMD getApplicableLanguages
FINE: Using PLSQL version: PLSQL
lip 23, 2019 2:56:17 PM net.sourceforge.pmd.PMD processFiles
WARNING: This analysis could be faster, please consider using Incremental Analysis: https://pmd.github.io/latest/pmd_userdocs_incremental_analysis.html
lip 23, 2019 2:56:17 PM net.sourceforge.pmd.RulesetsFactoryUtils printRuleNamesInDebug
FINER: Loaded rule AvoidTabCharacter
lip 23, 2019 2:56:17 PM net.sourceforge.pmd.processor.PmdRunnable call
FINE: Processing C:\Users\...\01_DB\00_Scripts\sols46_com_group_agent.sql
lip 23, 2019 2:56:17 PM net.sourceforge.pmd.processor.PmdRunnable addError
FINE: Error while processing file: C:\Users\...\01_DB\00_Scripts\sols46_com_group_agent.sql
net.sourceforge.pmd.PMDException: Error while parsing C:\Users\...\01_DB\00_Scripts\sols46_com_group_agent.sql
        at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:99)
        at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:51)
        at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:78)
        at net.sourceforge.pmd.processor.PmdRunnable.call(PmdRunnable.java:24)
        at java.util.concurrent.FutureTask.run(Unknown Source)
        at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at java.util.concurrent.FutureTask.run(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: net.sourceforge.pmd.lang.plsql.ast.ParseException: Encountered " "." ". "" at line 8, column 28.
Was expecting one of:
    "," ...
    "FROM" ...
    "," ...
    "FROM" ...

        at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.generateParseException(PLSQLParser.java)
        at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.jj_consume_token(PLSQLParser.java)
        at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.FromClause(PLSQLParser.java:20506)
        at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.RestOfStatement(PLSQLParser.java:10592)
        at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.SelectStatement(PLSQLParser.java:27804)
        at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.UnlabelledStatement(PLSQLParser.java:22950)
        at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Statement(PLSQLParser.java:23875)
        at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Block(PLSQLParser.java:952)
        at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Global(PLSQLParser.java:505)
        at net.sourceforge.pmd.lang.plsql.ast.PLSQLParser.Input(PLSQLParser.java:202)
        at net.sourceforge.pmd.lang.plsql.PLSQLParser.parse(PLSQLParser.java:55)
        at net.sourceforge.pmd.SourceCodeProcessor.parse(SourceCodeProcessor.java:111)
        at net.sourceforge.pmd.SourceCodeProcessor.processSource(SourceCodeProcessor.java:175)
        at net.sourceforge.pmd.SourceCodeProcessor.processSourceCode(SourceCodeProcessor.java:96)
        ... 9 more
```

**Code Sample demonstrating the issue:**

```
begin

select id_no into t4666.rec.agent_no from name where company_reg_no = '66666111';

end;
/

```

**Running PMD through:** *[CLI]*

<!-- If relevant, also include your JDK and OS information, e.g. for ClassNotFoundException, LinkageError, reflection failures, etc. -->

--- a/pmd-plsql/etc/grammar/PldocAST.jjt
+++ b/pmd-plsql/etc/grammar/PldocAST.jjt
@@ -1885,6 +1885,7 @@ ASTVariableName VariableName() :
 {
     id = ID() { name.append(id.getImage()); }
     [ "." id = ID() { name.append('.').append(id.getImage()); } ]
+    [ "." id = ID() { name.append('.').append(id.getImage()); } ]
     {
         jjtThis.setImage(name.toString());
         return jjtThis;
@@ -2325,7 +2326,7 @@ ASTValuesClause ValuesClause() :
     (
         "(" ( Expression() | <_DEFAULT> ) ( "," ( Expression() | <_DEFAULT> ) )* ")"
     |
-        ID() // that's a record variable name
+        Name() // that's a record variable name
     )
     { return jjtThis; }
 }
@@ -2409,7 +2410,7 @@ ASTUpdateSetClause UpdateSetClause() :
 {
     <SET>
     (
-        LOOKAHEAD(1) <ROW> "=" ID()
+        LOOKAHEAD(1) <ROW> "=" Name()
     |
         "VALUE" "(" TableAlias() ")" "=" ( LOOKAHEAD(1) "(" Subquery() ")" | Expression() )
     |
@@ -2431,7 +2432,7 @@ ASTUpdateSetClause UpdateSetClause() :
 ASTReturningClause ReturningClause() :
 {}
 {
-    ( <RETURN> | <RETURNING> ) ( Expression() (",")? )+ <INTO> ( ID() (",")? )+
+    ( <RETURN> | <RETURNING> ) ( Expression() (",")? )+ <INTO> ( Name() (",")? )+
     { return jjtThis; }
 }
 

--- a/pmd-plsql/src/test/java/net/sourceforge/pmd/lang/plsql/ast/RecordTypeTest.java
+++ b/pmd-plsql/src/test/java/net/sourceforge/pmd/lang/plsql/ast/RecordTypeTest.java
@@ -1,0 +1,24 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.plsql.ast;
+
+import java.nio.charset.StandardCharsets;
+
+import org.apache.commons.io.IOUtils;
+import org.junit.Assert;
+import org.junit.Test;
+
+import net.sourceforge.pmd.lang.plsql.AbstractPLSQLParserTst;
+
+public class RecordTypeTest extends AbstractPLSQLParserTst {
+
+    @Test
+    public void parseRecordType() throws Exception {
+        String code = IOUtils.toString(this.getClass().getResourceAsStream("RecordType.pls"),
+                StandardCharsets.UTF_8);
+        ASTInput input = parsePLSQL(code);
+        Assert.assertNotNull(input);
+    }
+}

--- a/pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/RecordType.pls
+++ b/pmd-plsql/src/test/resources/net/sourceforge/pmd/lang/plsql/ast/RecordType.pls
@@ -1,0 +1,47 @@
+create or replace package body solt9001 is
+
+  procedure upd is
+
+  begin
+
+      update solt90_web_service_log
+      set    row = solt9001.rec
+      returning record_version into solt9001.rec.record_version;
+
+  end upd;
+
+end;
+/
+
+create or replace package body solt9001 is
+
+  procedure upd is
+
+  begin
+
+      update solt90_web_service_log
+      set    row = solt9001.rec
+      returning record_version into solt9001.rec.record_version;
+
+  end upd;
+
+end;
+/
+
+create or replace function test return varchar2 is
+  begin
+
+    insert into sol_individual_commission_rate
+    values solt4601.rec
+    returning record_version, rowid into solt4601.rec.record_version, solt4601.current_rowid;
+
+    return null;
+  end test;
+/
+
+begin
+
+select id_no into t4666.rec.agent_no from name where company_reg_no = '66666111';
+
+end;
+/

